Jinacore-496 add docs for remote jina update (#1675)
jina-ai/jina: docs/chapters/remote/jinad.md

# Jina Daemon

`jinad`, aka `Jina Daemon`, is a persistent process for deploying and managing Jina Flow, Pods, and Peas in a distributed system. 

## Terminologies

- *Workflow*: a set of connected pods for accomplishing certain task, e.g. indexing, searching, extracting.
- *Flow API*: a pythonic way for users to construct workflows in Jina with clean, readable idioms. 
- *Remote*, *remote instance*, *remote machine*: the place where you want to run the pod, the place offers better computational capability or larger storage. For example, one may want to run an encode pod on the remote GPU instance.  
- *Local*, *local instance*, *local machine*: the place of your entrypoint and the rest parts of your workflow.

## Design

`jinad` is designed to maintain bookkeeping for the running Flows, Pods and Peas on the remote machines. `jinad` can also spawn Pods and Peas to other remote machines that have `jinad` running. [`fluentd`](https://github.com/fluent/fluentd) is used to collect logs from different Processes and ensure the logs belonging to the same Flow are stored consistently. 

![jinad design](jinad_design.png)

## Installation

### Using Docker Images (Recommended)

> The simplest way to use `jinad` is via Docker. There is no need to worry about fluent installation. You only need to have [Docker installed](https://docs.docker.com/install/) first. 

Reviewer: fluent -> fluentd

```bash
pip install "git+https://github.com/jina-ai/jina.git#egg=jina[daemon]"
```

### Install from Your Local Fork/Clone

If you are a developer and want to test your changes on-the-fly: 

```bash
git clone https://github.com/jina-ai/jina
cd jina && pip install -e ".[daemon]"
``` 

## Usage 

### Prerequisites
Run `jinad` on the remote machine. We assume the remote is in the intranet and its IP address is `12.34.56.78`. By default, `jinad` will use the port `8000` for receiving requests. Make sure `8000` port is publicly accessible.

After having `jinad` running on the remote, you can open the browser the visit `http://3.16.166.3:8000/alive` to check whether `jinad` is properly set up. If everythong works well, you will see the following response.

```json
{"status_code":200,"jina_version":"0.9.12"}
```

#### Using Docker Container

We start a Docker container under the `host` mode so that it will connect all the ports to the host machine. `-d` option is to keep the container running in the background

```bash
docker run -d --network host jinaai/jina:latest-daemon
```

#### Using Native Python

```bash
jinad
```

### Creating a Remote Pod from Console

> In the simplest case, you may want to create a Pod on the remote. The most naive way is to log into the remote machine and start a pod using `jina` CLI [link to run-remote-pod-console](). To avoid logging into the remote machine every time, we can use `jinad` to do the same thing. Furthermore, `jinad` offers a session management for the running Pods on the remote and saves you from manually creating and deleting Pods.

Reviewer: This link will be updated later, right?

           Flow@69676[S]:🎉 Flow is ready to use, accepting gRPC request
           Flow@69676[I]:
	🖥️ Local access:	tcp://0.0.0.0:51327
	🔒 Private network:	tcp://192.168.1.9:51327
	🌐 Public address:	tcp://203.184.132.69:51327
         Client@69676[S]:connected to the gateway at 0.0.0.0:51327!
...
        gateway@69697[I]:send: 1 recv: 0 pending: 1
pod0/ZEDRuntime@69694[I]:recv IndexRequest  from gateway▸pod0/ZEDRuntime▸⚐
pod0/ZEDRuntime@69694[I]:#sent: 0 #recv: 1 sent_size: 0 Bytes recv_size: 423 Bytes
🌏 pod1/ZEDRuntime@69696[I]:recv IndexRequest  from gateway▸pod0/ZEDRuntime▸pod1/ZEDRuntime▸⚐
🌏 pod1/ZEDRuntime@69696[I]:#sent: 0 #recv: 1 sent_size: 0 Bytes recv_size: 508 Bytes
...
	✅ done in ⏱ 2 seconds 🐎 47.9/s
        gateway@69676[S]:terminated
...
           pod1@69676[S]:terminated
...
           pod0@69676[S]:terminated
...
```

### Creating a Remote Flow
One common use case in production is running the whole Flow on the remote with pods distributed on different machines. In this case, the very first thing is to ensure all the remote machines have `jinad` running properly. We will see how to create a remote Flow from the local machine.

#### 1. Create a Flow 

> As `jinad` hosts a service on the remote, we can use the `jinad` API `/flow/yaml` to create a Flow via uploading a Flow configuration yaml file. Please refere to the full  [API specifications](https://api.jina.ai/daemon/) for more details.

Reviewer: typo (refer)

> As `jinad` hosts a service on the remote, we can use the `jinad` API `/flow/yaml` to create a Flow via uploading a Flow configuration yaml file. Please refere to the full  [API specifications](https://api.jina.ai/daemon/) for more details.

Reviewer: ```suggestion
As `jinad` hosts a service on the remote, we can use the `jinad` API `/flow/yaml` to create a Flow via uploading a Flow configuration yaml file. Please refer to the full  [API specifications](https://api.jina.ai/daemon/) for more details.
```

```python
import requests

def create_flow(flow_url, yamlspec):
    flow_creation_api = f'{flow_url}/flow/yaml/'
    with open(yamlspec, 'rb') as f:
        files = [('yamlspec', f)]
        try:
            r = requests.put(url=flow_creation_api, files=files, timeout=10)
            if r.status_code == requests.codes.ok:
                return r.json()['flow_id']
            else:
                print('Remote Flow creation failed')
        except requests.exceptions.RequestException as ex:
            print(f'Exception raised: {ex!r}')

if __name__ == '__main__':
    host_ip = '3.16.166.3'
    jinad_port = '8000'
    flow_api = f'http://{host_ip}:{jinad_port}'
    flow_id = create_flow(flow_api, 'dummy_flow.yml')
    print(f'flow is created: {flow_id}')
```

We use a dummy Flow configuration `dummy_flow.yml` as below for demo purpose.

```yaml
!Flow
version: 1
pods:
  - name: pod0
    method: add
    uses: _logforward
  - name: pod1
    method: add
    uses: _logforward
```

> After running the above codes, we print out the flow id and please write down this flow id for the next step. 

Reviewer: ```suggestion
After running the above codes, we print out the flow id and please note this flow id for the next step. 
```

```text
flow is created: cdd53e16-5575-11eb-86b2-0ab9db700358
```

> #### 2. Check Flow StatusFlow 

Reviewer 2: ```suggestion
#### 2. Check Flow Status
```

> Before using the Flow we created, we need to get detail information about the Flow so that we can send queries to the Flow. To get the information of the Flow, we can use the following codes to retrieve the information from `/flow/cdd53e16-5575-11eb-86b2-0ab9db700358`

Reviewer: I'm unsure of the usage of the word `codes` here. Would `script` be better?

Author: sounds better.

            return str(r.json()["port"])
        else:
            print('Remote Flow info retrieval failed')
    except requests.exceptions.RequestException as ex:
        print(f'Exception raised: {ex!r}')

def main():
    host_ip = '3.16.166.3'
    jinad_port = '8000'
    flow_api = f'http://{host_ip}:{jinad_port}'
    flow_id = 'cdd53e16-5575-11eb-86b2-0ab9db700358'
    flow_port = get_flow_info(flow_api, flow_id)
    print(f'Flow serves at {host_ip}:{flow_port}')
``` 

#### 3. Run Index/Query via client 
With the host and port information of the remote Flow, we can use the gRPC client of jina to send index or query requests.

```python
from jina.parsers import set_client_cli_parser
from jina.clients import Client
from jina import Document

def send_index_request(host, port):
    print(f'index request sent to {host}:{port}')
    args = set_client_cli_parser().parse_args(
        ['--host', host, '--port-expose', str(port)])
    grpc_client = Client(args)
    grpc_client.index(
        [Document(text='hello, jina'), ], on_done=print)

def main():
    host_ip = '3.16.166.3'
    flow_port = '51871'
    send_index_request(host_ip, flow_port)
``` 

> #### 4. Close Flow

Reviewer 2: ```suggestion
#### 4. Terminate Flow
```
